garage/tf/core/parameterLayer.py

```diff
@@ -0,0 +1,54 @@
+"""Parameter layer in TensorFlow."""
+import tensorflow as tf
+from tensorflow.keras.layers import Layer as KerasLayer
+from tensorflow.python.ops.gen_array_ops import broadcast_to
+
+
+class ParameterLayer(KerasLayer):
+    """
+    Parameter layer based on tf.keras.layers.Layer.
+
+    Used as layer that could be broadcast to a certain shape to
+    match with input variable during training.
+    Example: A trainable parameter variable with shape (2,), it needs to be
+    broadcasted to (32, 2) when applied to a batch with size 32.
+    """
+
+    def __init__(self,
+                 length,
+                 scope='ParameterLayer',
+                 initializer='ones',
+                 trainable=True,
+                 **kwargs):
+        self.length = length
+        self.initializer = initializer
+        self.trainable = trainable
+        self.scope = scope
+        super().__init__(**kwargs)
+
+    def build(self, input_shape):
+        """tf.keras.layers.Layer build."""
+        with tf.variable_scope(self.scope):
```

Comment:
how does Keras handle scope? I wonder, does it handle scope automatically?

Reply:
TensorFlow handles variable handling by variable scope, i.e. if two variables with same name are declared under the same variable, they share the same variable/weight. 

Keras handles variable handling in a different way. Variable sharing should be done via calling a same Keras layer (or model) instance multiple times. **A TensorFlow variable scope will have no effect on a Keras layer or model**. (Reference: [Keras](https://blog.keras.io/keras-as-a-simplified-interface-to-tensorflow-tutorial.html))

I am not completely positive whether tf.Keras follows this, but for sure we can use `tf.name_scope` for better graph visualization.

Reply:
No, tf.keras doesn't handle scope automatically. Because a tf.keras.Layer will generate it's [unique_layer_name](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/base_layer.py#L1257), so if we want it to handle the scope, we have pass it to every layers. But it seems in 2.x they will fix [this](https://github.com/tensorflow/tensorflow/issues/24079).

Reply:
so looks like we can use `tf.variable_scope` for sharing variables, and pass the name to the layers for now?

Reply:
I think so but you need to test it. 
FYI, I think using name_scope here doesn't seem to work:
``` python
model = GaussianMLPModel(
            input_dim=5, output_dim=2, hidden_sizes=(4, 4), init_std=2.0, adaptive_std=True)
```
error:
```
ERROR: test_autopickable_gaussian_mlp_pickling (tests.garage.tf.models.test_model_auto_pickable.TestKerasModel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/chang/Code/garage/tests/garage/tf/models/test_model_auto_pickable.py", line 93, in test_autopickable_gaussian_mlp_pickling
    input_dim=5, output_dim=2, hidden_sizes=(4, 4), init_std=2.0, adaptive_std=True)
  File "/home/chang/Code/garage/garage/tf/models/gaussian_mlp_model.py", line 98, in __init__
    self.model = self.build_model()
  File "/home/chang/Code/garage/garage/tf/models/gaussian_mlp_model.py", line 189, in build_model
    outputs=[mean_var, std_var, std_param_var, sample_var])
  File "/home/chang/miniconda2/envs/garage/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 111, in __init__
    super(Model, self).__init__(*args, **kwargs)
  File "/home/chang/miniconda2/envs/garage/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 78, in __init__
    self._init_graph_network(*args, **kwargs)
  File "/home/chang/miniconda2/envs/garage/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 244, in _init_graph_network
    self.inputs, self.outputs)
  File "/home/chang/miniconda2/envs/garage/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 1714, in _map_graph_network
    str(layers_with_complete_input))
ValueError: Graph disconnected: cannot obtain value for tensor Tensor("GaussianMLPModel/input_2:0", shape=(?, 5), dtype=float32) at layer "input_2". The following previous layers were accessed without issue: []

----------------------------------------------------------------------
Ran 5 tests in 2.418s

```

Reply:
tf.keras.Model and tf.keras.Layer don't handle variable_scope, because the name of the layer is assigned by passing a name value when constructing the object.
When constructing a tf.keras.Model, it will check all the layer names inside the model. If there is a hidden_1 in mean_network, another hidden_1 in std_network, the summary of the GaussianMLPModel would become:
```
Layer (type)        Output Shape         Param #     Connected to                            
__________________________________________________________________________________________________
hidden_1 (Dense)    (None, 4)            20          hidden_0[0][0]                    
__________________________________________________________________________________________________
hidden_1 (Dense)    (None, 4)            20          hidden_0[0][0]       

```
This is invalid because there are two `hidden_1` layers.
Based on the above test,we may as well conclude that they can't share the same variable by using `with tf.variable_scope("variable_scope", reuse=True)`.

Reply:
Right. `tf.variable_scope` is simply useless here. I tried with appending the scope to the variables name in the latest commit. It looks pretty good like this
<img width="1006" alt="screen shot 2019-01-29 at 7 05 37 pm" src="https://user-images.githubusercontent.com/10270515/51955481-e4dc1180-23f8-11e9-806c-b53e81bf1c43.png">
where there is one GaussianMLP with `adaptive_std` and one doesn't, and they share the input placeholder.

Reply:
i think in this case variable sharing should be a non-issue, because models will allow us to abandon re-building the whole network over-and-over again like rllab did.

for example, GaussianMLPPolicy.entropy_sym() rebuilds the entire network (sharing weights with the original) and then returns the entropy variable. A GaussianMLPPolicy written with Models should just have a call entropy() which returns the the entropy tensor op of the underlying distribution.